### PR TITLE
[ISSUE #16] Implemented a function and tests to extract vertices from spoly by index

### PR DESCRIPTION
**Pull request by stepan-neretin7**
@esabol @vitcpp please, review it.

### CODE REVIEW

File: sql/poly.sql [hunk 1]
```diff
@@ -593,3 +593,13 @@ SELECT npoints( spoly '{
...
+
+SELECT set_sphere_output( 'RAD' );
+
+SELECT  spoint( spoly '{(0,0),(1,0),(1,1)}', 1 );
+
+SELECT  spoint( spoly '{(0,0),(1,0),(1,1)}', 2 );
+
+SELECT  spoint( spoly '{(0,0),(1,0),(1,1)}', 3 );
+
+SELECT  spoly_as_array( spoly '{(0,0),(1,0),(1,1)}' );
```
Comment (esabol) [inline sql/poly.sql:605]: I hate to nitpick whitespace, but I recommend removing one of the two spaces after the SELECTs on lines 599, 601, 603, and 605.

Otherwise, it looks good to me! But if you're adding a function, shouldn't there be an upgrade script and version number bump? Or will that be done in a separate PR?

**Review by esabol — CHANGES_REQUESTED** [review]

File: sql/poly.sql (hunk 1 above)
Comment (stepan-neretin7) [reply sql/poly.sql:605]: Of course, but which version should we put? @vitcpp

Comment (vitcpp) [reply sql/poly.sql:605]: As we discussed before, the master branch is the development branch. Once the API is changed I propose to assign a new version 1.3.0. Upgrade script should be named as pg_sphere--1.2.3--1.3.0.sql.in. The same as I proposed in #22.

Comment (esabol) [reply sql/poly.sql:602]: Except PR #22 already bumps the version to 1.3.0 and adds an upgrade script. I propose merging PR #22 first and then git rebase this PR on the new master to add this functionality to the future 1.3.0.

File: Makefile [hunk 2]
```diff
@@ -260,6 +261,12 @@ endif
 pg_sphere--1.2.2--1.2.3.sql:
 	cat upgrade_scripts/$@.in > $@
 
+<<<<<<< HEAD
```
Comment (esabol) [inline Makefile]: You have a merge conflict here.

Comment (stepan-neretin7) [reply Makefile]: fixed

**Comment by stepan-neretin7** [comment]
What about this pull request? Is it approved? @esabol

**Comment by esabol** [comment]
> What about this pull request? Is it approved? @esabol

It's perfect other than PR https://github.com/postgrespro/pgsphere/pull/38 also bumps the version to 1.3.0 and adds an upgrade script. So I recommended merging PR https://github.com/postgrespro/pgsphere/pull/38 first and then `git rebase` this PR on the new master. Or you could do it in the opposite order, if you prefer.

Thank you, @stepan-neretin7, for all your work!

**Comment by stepan-neretin7** [comment]
Thank you also very much for your patient, detailed review of my code (I'm just learning and it's very valuable to me). Yes, I have added a version update here too and there will be no problems with rebase in any order. Therefore, @vitcpp can merge in any order
It turns out that this pull request is also approved by @esabol?

**Comment by esabol** [comment]
@stepan-neretin7 : Now that PR #38 has been merged, I think this PR could use a `git rebase` since the changes to the Makefile are no longer needed and the changes to the upgrade script will likely conflict.

**Review by esabol — APPROVED** [review]
Looks good to me! Thank you, @stepan-neretin7

**Comment by stepan-neretin7** [comment]
Can we merge it? @vitcpp

File: src/polygon.h [hunk 3]
```diff
@@ -72,6 +72,11 @@ bool	spoly_segment(SLine *sl, const SPOLY *poly, int32 i);
  */
 bool	spoly_contains_point(const SPOLY *pg, const SPoint *sp);
 
+/*
+ * Returns the n-th point of a spoly.
+*/
```
Comment (vitcpp) [inline src/polygon.h:77]: Fix identation.

File: src/polygon.h [hunk 4]
```diff
@@ -348,4 +353,9 @@ Datum	spherepoly_add_point(PG_FUNCTION_ARGS);
  */
 Datum	spherepoly_add_points_finalize(PG_FUNCTION_ARGS);
 
+/*
+* Returns spoly as array of points
```
Comment (vitcpp) [inline src/polygon.h:357]: FIx identation

File: src/polygon.c [hunk 5]
```diff
@@ -556,6 +558,58 @@ spoly_segment(SLine *sl, const SPOLY *poly, int32 i)
...
+}
+
+Datum
+spherepoly_get_array(PG_FUNCTION_ARGS)
+{
+    SPOLY *poly = PG_GETARG_SPOLY(0);
+    Datum *datum_arr = (Datum *) palloc(sizeof(Datum) * poly->npts);
+    ArrayType *res;
+	SPoint *p = (SPoint *) palloc(sizeof(SPoint) * poly->npts);
+    for (size_t i = 0; i < poly->npts; i++)
```
Comment (vitcpp) [inline src/polygon.c:595]: Add blank line above for.

File: src/polygon.c [hunk 6]
```diff
@@ -556,6 +558,58 @@ spoly_segment(SLine *sl, const SPOLY *poly, int32 i)
...
+	PG_RETURN_NULL();
+}
+
+Datum
+spherepoly_get_array(PG_FUNCTION_ARGS)
+{
+    SPOLY *poly = PG_GETARG_SPOLY(0);
+    Datum *datum_arr = (Datum *) palloc(sizeof(Datum) * poly->npts);
+    ArrayType *res;
+	SPoint *p = (SPoint *) palloc(sizeof(SPoint) * poly->npts);
```
Comment (vitcpp) [inline src/polygon.c:594]: Fix identation

File: src/point.h [hunk 7]
```diff
@@ -3,6 +3,8 @@
 
 #include "vector3d.h"
 #include "sbuffer.h"
+#include <catalog/namespace.h>
+#include <utils/resowner.h>
```
Comment (vitcpp) [inline src/point.h:7]: Why do we need to include these headers?

File: src/path.h [hunk 8]
```diff
@@ -71,6 +71,11 @@ Datum	spherepath_in(PG_FUNCTION_ARGS);
  */
 Datum	spherepath_get_point(PG_FUNCTION_ARGS);
 
+/*
+* Returns spath as array of points
```
Comment (vitcpp) [inline src/path.h:75]: FIx identation

**Review by vitcpp — CHANGES_REQUESTED** [review]

File: src/point.h (hunk 7 above)
Comment (stepan-neretin7) [reply src/point.h]: #include <catalog/namespace.h> only needed

File: src/point.h [hunk 9]
```diff
@@ -15,6 +17,11 @@ typedef struct
 	float8	lat;	/* latitude value in radians */
 } SPoint;
 
+__attribute__((unused))
```
[the remark this answers is not in the document]
Comment (stepan-neretin7) [reply src/point.h:19]: really required for disable unused  warning in other included places

**Review by vitcpp — CHANGES_REQUESTED** [review]

File: src/path.h (hunk 8 above)
Comment (vitcpp) [inline src/path.h:75]: Fix identation

image

File: src/path.c [hunk 10]
```diff
@@ -555,6 +558,32 @@ spherepath_get_point(PG_FUNCTION_ARGS)
...
+    Datum *datum_arr = (Datum *) palloc(sizeof(Datum) * path->npts);
+    ArrayType *res;
+	SPoint *p = (SPoint *) palloc(sizeof(SPoint) * path->npts);
+
+    for (size_t i = 0; i < path->npts; i++)
+    {
+        if (!spath_get_point(&p[i], path, i))
+        {
+            // Clean up and return NULL
+            for (size_t j = 0; j < i; j++)
```
Comment (vitcpp) [inline src/path.c:574]: Invalid memory free. Remove for loop. Just use pfree(p). It should be done like in spherepoly_get_array.

File: src/polygon.c [hunk 11]
```diff
@@ -556,6 +558,57 @@ spoly_segment(SLine *sl, const SPOLY *poly, int32 i)
 	}
 }
 
+static bool
+poly_get_point(SPoint *sp, const SPOLY *poly, int32 i)
```
Comment (vitcpp) [inline src/polygon.c:562]: I propose to rename the function name to spoly_get_point to be consistent with the rest of the code. spoly prefix is used in names of other functions.

**Review by esabol — APPROVED** [review]

**Review by vitcpp — APPROVED** [review]
